### PR TITLE
Funnel all "get local conformances" queries through a request.

### PR DESCRIPTION
**Pull request by DougGregor**
`LookupAllConformancesInContextRequest` caches the results of performing
lookup of all of the conformances ascribed to a given iterable
declaration context. However, it was only used in a small number of
places, with most clients using a different API
(`getLocalConformances()`) that does not provide
caching, cycle detection, or dependency tracking.

Sink `LookupAllConformancesInContextRequest` lower in the stack, and
reimplement `getLocalConformances()` on top of it. This ensures that
all of the various queries go through the cached request and get the
benefits of the request-evaluator infrastructure.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/ProtocolConformance.cpp [hunk 1]
```diff
@@ -1321,59 +1321,83 @@ NominalTypeDecl::getSatisfiedProtocolRequirementsForMember(
...
+    const {
+  // Look up the cached set of all of the conformances.
+  std::vector<ProtocolConformance *> conformances =
+      evaluateOrDefault(
+        getASTContext().evaluator, LookupAllConformancesInContextRequest{this},
+        { });
+
+  // Copy all of the conformances we want.
+  SmallVector<ProtocolConformance *, 2> result;
+  std::copy_if(
```
Comment (slavapestov) [inline lib/AST/ProtocolConformance.cpp:1371]: You could bypass the copy_if() if the lookupKind is All, which is the default here.

Comment (DougGregor) [reply lib/AST/ProtocolConformance.cpp:1371]: I have a silly SmallVector-vs-vector difference to deal with as well, but yes, thanks!

**Pull request merged**